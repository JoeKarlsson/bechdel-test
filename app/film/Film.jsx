import React from 'react';
import * as $ from'jquery';
import { Link } from 'react-router';

export default React.createClass({
  getInitialState() {
    return {
      film: {}
    }
  },

  getFilm() {
    $.ajax({
      url: "/api/film/" + this.props.params.id,
      method: 'GET',
      dataType: 'json',
      cache: false,
      success: function(data) {
        this.setState({ film: data });
        console.log(this.state.film, 'data')
      }.bind(this),
      error: function(xhr, status, err) {
        console.error(this.props, status, err.toString());
      }.bind(this)
    });
  },

  componentDidMount() {
    this.getFilm();
  },

  render() {
    return (
      <div>
        <h2>{this.state.film.title}</h2>
        <p>{this.state.film.plot}</p>
        <Link to={'/films'}><button>All Films</button></Link>
      </div>
    )
  }
});
