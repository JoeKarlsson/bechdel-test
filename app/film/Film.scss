.films {
  background: pink;
}

.filmListContainer {
  background: blue;
  display: flex;
}