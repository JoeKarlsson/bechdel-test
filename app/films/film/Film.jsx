import React from 'react';
import { Link } from 'react-router';
import styles from './Film.scss';
import * as $ from 'jquery';
import Chart from '../../graphs/Chart.jsx';

class Film extends React.Component {
  constructor() {
    super();
    this.state = {
      film: {
        title: '',
        images: {
          poster: '',
          backdrop: '',
        },
        plot: '',
        directors: [],
        writers: [],
        genres: [],
        rated: '',
        bechdelResults: {
          pass: 'false',
          bechdelScore: 0,
          numScenesPass: 0,
          scenesThatPass: [],
          numScenesDontPass: 0,
          numOfFemalesChars: 0,
          numOfMaleChars: 0,
          numOfFemalesCharsWithDialogue: 0,
          numOfMaleCharsWithDialogue: 0,
          totalLinesFemaleDialogue: 0,
          totalLinesMaleDialogue: 0,
        },
      },
    };
    this.getFilm = this.getFilm.bind(this)
  }

  getFilm() {
    $.ajax({
      url: "/api/film/" + this.props.params.id,
      method: 'GET',
      dataType: 'json',
      cache: false,
      success: function(data) {
        this.setState({ film: data });
      }.bind(this),
      error: function(xhr, status, err) {
        console.error(this.props, status, err.toString());
      }.bind(this)
    });
  }

  componentDidMount() {
    this.getFilm();
  }

  render() {
    const directorNode = this.state.film.directors.map((director) => {
      const directors = this.state.film.directors;
      if ( director.name === directors[directors.length-1].name) {
        return (
          <span key={director.id}>{director.name}</span>
        );
      }
      return (
        <span key={director.id}>{director.name} & </span>
      );
    });
    const writerNode = this.state.film.writers.map((writer) => {
      const writers = this.state.film.writers;
      if ( writer.name !== writers[writers.length-1].name) {
        return (
          <span key={writer.id}>{writer.name} & </span>
        );
      } else {
        return (
          <span key={writer.id}>{writer.name}</span>
        );
      }
    });
    const genreNode = this.state.film.genres.map((genre) => {
      const genres = this.state.film.genres;
      if ( genre !== genres[genres.length-1]) {
        return (
          <span key={Math.floor(Math.random()*100)}>{genre} | </span>
        );
      } else {
        return (
          <span key={Math.floor(Math.random()*100)}>{genre}</span>
        );
      }
    });
    return (
      <div>
        <h1>{this.state.film.title}</h1>
        <h3>Bechdel Pass: {this.state.film.bechdelResults.pass.toString().toUpperCase()}</h3>
        <p>Bechdel Score: {this.state.film.bechdelResults.bechdelScore} of 3</p>
        <img
          src={this.state.film.images.backdrop}
          alt={this.state.film.title}
        />
        <p>Directors: {directorNode}</p>
        <p>Writers: {writerNode}</p>
        <p>Genre: {genreNode}</p>
        <p>Rated: {this.state.film.rated}</p>
        <p>IMDB: <a href={`http://www.imdb.com/title/${this.state.film.idIMDB}`} target='_blank'>{this.state.film.title}</a></p>
        <p>Bechdel Score: {this.state.film.bechdelResults.bechdelScore} of 3</p>
        <p>Number of Scenes that pass: {this.state.film.bechdelResults.numScenesPass}</p>
        <p>Number of Scenes that dont pass: {this.state.film.bechdelResults.numScenesDontPass}</p>
        <p>Number Of Females Characters: {this.state.film.bechdelResults.numOfFemalesChars}</p>
        <p>Number Of Male Characters: {this.state.film.bechdelResults.numOfMaleChars}</p>
        <p>Number of Females Characters With Dialogue: {this.state.film.bechdelResults.numOfFemalesCharsWithDialogue}</p>
        <p>Number of Male Characters With Dialogue: {this.state.film.bechdelResults.numOfMaleCharsWithDialogue}</p>
        <p>Total Lines of Female Dialogue: {this.state.film.bechdelResults.totalLinesFemaleDialogue}</p>
        <p>Total Lines of Male Dialogue: {this.state.film.bechdelResults.totalLinesMaleDialogue}</p>
        <p>{this.state.film.plot}</p>

        <Chart />
        <Link to={'/'}><button>All Films</button></Link>
      </div>
    )
  }
}

Film.propTypes = {
  film: React.PropTypes.object,
};

export default Film;
