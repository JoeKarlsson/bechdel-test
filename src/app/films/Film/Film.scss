.filmInfo {

	.results {
		margin: 2%;
		display: flex;
		justify-content: space-around;

		.catName {
			font-weight: bold;
		}
	}
}

.filmListContainer {
	background: blue;
	display: flex;
}
