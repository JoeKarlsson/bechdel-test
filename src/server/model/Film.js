/* eslint
  func-names: 0,
  arrow-body-style: 0
*/

require('./index.js');
const mongoose = require('mongoose');
const parseData = require('./parseData');

mongoose.Promise = global.Promise;

const filmSchema = mongoose.Schema({
	title: {
		type: String,
		required: true,
	},
	plot: String,
	simplePlot: String,
	year: Number,
	releaseDate: String,
	actors: [
		{
			actorName: String,
			character: String,
			actorActress: String,
		},
	],
	directors: [
		{
			name: String,
			id: String,
		},
	],
	writers: [
		{
			name: String,
			id: String,
		},
	],
	rated: String,
	genres: [String],
	urlPoster: String,
	idIMDB: String,
	urlIMDB: String,
	rating: String,
	metascore: Number,
	createdAt: {
		type: Date,
		default: Date.now,
	},
	bechdelResults: {
		pass: Boolean,
		bechdelScore: Number,
		numScenesPass: Number,
		numScenesDontPass: Number,
		numOfFemalesChars: Number,
		numOfMaleChars: Number,
		numOfFemalesCharsWithDialogue: Number,
		numOfMaleCharsWithDialogue: Number,
		totalLinesFemaleDialogue: Number,
		totalLinesMaleDialogue: Number,
	},
	images: {
		backdrop: String,
		poster: String,
	},
});

filmSchema.static('listAll', function() {
	const promise = new Promise(resolve => {
		this.find()
			.sort('-date')
			.exec()
			.then(result => {
				return resolve(result);
			});
	});
	return promise;
});

filmSchema.static('findByID', function(id) {
	const promise = new Promise((resolve, reject) => {
		this.find({ _id: id })
			.sort('-date')
			.exec()
			.then(result => {
				if (Array.isArray(result)) {
					return resolve(result[0]);
				}
				return resolve(result);
			})
			.catch(err => {
				reject(new Error(err));
			});
	});
	return promise;
});

filmSchema.static('findByTitle', function(title) {
	const promise = new Promise((resolve, reject) => {
		if (!title) {
			reject(new Error('No film tile found'));
		}
		this.find({ title })
			.exec()
			.then(result => {
				if (result.length === 0) {
					reject(new Error('film not found'));
				}
				if (Array.isArray(result)) {
					return resolve(result[0]);
				}
				return resolve(result);
			})
			.catch(err => {
				reject(new Error(err));
			});
	});
	return promise;
});

filmSchema.static('deleteFilm', function(id) {
	if (!id) {
		return 'Invalid input on deleteFilm';
	}
	return this.findOne({ _id: id })
		.exec()
		.then(film => film.remove)
		.then(() => true)
		.catch(error => {
			throw new Error(error);
		});
});

filmSchema.static('insertFilm', filmMetaData => {
	const promise = new Promise((resolve, reject) => {
		const { title, bechdelResults, actors, images, data } = filmMetaData;
		const film = new Film({ title });
		film.title = title;
		film.bechdelResults = bechdelResults;
		film.plot = data.plot;
		film.simplePlot = data.simplePlot;
		film.year = data.year;
		film.releaseDate = data.releaseDate;
		film.directors = data.directors;
		film.writers = data.writers;
		film.rated = data.rated;
		film.genres = data.genres;
		film.urlPoster = data.urlPoster;
		film.idIMDB = data.idIMDB;
		film.rating = data.rating;
		film.metascore = data.metascore;
		film.urlIMDB = data.urlIMDB;
		film.actors = parseData.parseActorArr(actors);
		film.images = parseData.parseImageData(images);

		return film
			.save()
			.then(result => {
				return resolve(result);
			})
			.catch(err => {
				return reject(new Error(err));
			});
	});
	return promise;
});

const Film = mongoose.model('Film', filmSchema);

module.exports = Film;
