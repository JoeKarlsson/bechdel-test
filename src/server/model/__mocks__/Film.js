const mockListAllResponse = ['foo', 'bar'];
const mockFindByIDResponse = { title: 'Boyhood' };

class Film {
	constructor() {

		this.getData = jest.fn(() => {
			const p = new Promise((resolve) => {
				resolve(mockFindByIDResponse);
			});
			return p;
		});

		this.listAll = jest.fn(() => {
			const p = new Promise((resolve) => {
				resolve(mockListAllResponse);
			});
			return p;
		});

		this.findByID = jest.fn(() => {
			const p = new Promise((resolve) => {
				resolve(mockFindByIDResponse);
			});
			return p;
		});

		this.deleteFilm = jest.fn(() => {
			const p = new Promise((resolve) => {
				resolve(mockFindByIDResponse);
			});
			return p;
		});

		this.findByTitle = jest.fn(() => {
			const p = new Promise((resolve) => {
				resolve(mockFindByIDResponse);
			});
			return p;
		});
	}
}

module.exports = new Film();
