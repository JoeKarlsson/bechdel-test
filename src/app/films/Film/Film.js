import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import ErrorBoundary from '../../shared/ErrorBoundary/ErrorBoundary';
import BechdelResults from './BechdelResults/BechdelResults';
import getAllFilmData from '../../helper/getAllFilms';
import hash from '../../helper/hash';
import './Film.scss';

class Film extends React.Component {
	constructor() {
		super();
		this.state = {
			film: {
				title: '',
				images: {
					poster: '',
					backdrop: '',
				},
				plot: '',
				directors: [],
				writers: [],
				genres: [],
				rated: '',
				bechdelResults: {
					pass: 'false',
					bechdelScore: 0,
					numScenesPass: 0,
					scenesThatPass: [],
					numScenesDontPass: 0,
					numOfFemalesChars: 0,
					numOfMaleChars: 0,
					numOfFemalesCharsWithDialogue: 0,
					numOfMaleCharsWithDialogue: 0,
					totalLinesFemaleDialogue: 0,
					totalLinesMaleDialogue: 0,
				},
			},
		};
	}

	componentDidMount() {
		const { id } = this.props.match.params;

		getAllFilmData(id)
			.then((data) => {
				this.setState({
					film: data,
				});
			});
	}

	render() {
		const {
			directors,
			writers,
			genres,
			title,
			bechdelResults,
			images,
		} = this.state.film;

		const directorNode = directors.map((director) => {
			if (director.name === directors[directors.length - 1].name) {
				return <span key={director.id}>{director.name}</span>;
			}
			return <span key={director.id}>{director.name} & </span>;
		});
		const writerNode = writers.map((writer) => {
			if (writer.name !== writers[writers.length - 1].name) {
				return <span key={writer.id}>{writer.name} & </span>;
			}
			return <span key={writer.id}>{writer.name}</span>;
		});
		const genreNode = genres.map((genre) => {
			if (genre !== genres[genres.length - 1]) {
				return <span key={hash(genre)}>{genre} | </span>;
			}
			return <span key={hash(genre)}>{genre}</span>;
		});

		return (
			<div className="filmInfo">
				<ErrorBoundary>
					<h1>
						{title}
					</h1>
					<h3>
						Bechdel Pass:{' '}
						{bechdelResults.pass.toString().toUpperCase()}
					</h3>
					<p>
						Bechdel Score: {bechdelResults.bechdelScore} of 3
					</p>
					<img
						src={images.backdrop}
						alt={title}
					/>
					<div className="plot">
						<p>{this.state.film.plot}</p>
					</div>
					<span className="results">
						<div className="filmData">
							<span className="catName">Directors:</span> {directorNode}
							<br />
							<span className="catName">Writers:</span> {writerNode} <br />
							<span className="catName">Genre:</span> {genreNode}
							<br />
							<span className="catName">Rated:</span>{' '}
							{this.state.film.rated}
							<br />
							<span className="catName">IMDB:</span>{' '}
							<a
								href={`http://www.imdb.com/title/${this.state.film.idIMDB}`}
								target="_blank"
							>
								{this.state.film.title}
							</a>
							<br />
						</div>
						<BechdelResults bechdelResults={bechdelResults} />
					</span>
					<Link to="/">
						<button>All Films</button>
					</Link>
				</ErrorBoundary>
			</div>
		);
	}
}

Film.propTypes = {
	match: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
};

Film.defaultProps = {};

export default Film;
