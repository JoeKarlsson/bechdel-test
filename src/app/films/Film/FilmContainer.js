import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Film from './Film';
import api from '../../helper/api';

class FilmContainer extends Component {
	constructor() {
		super();
		this.state = {
			film: {
				title: '',
				images: {
					poster: '',
					backdrop: '',
				},
				plot: '',
				directors: [],
				writers: [],
				genres: [],
				rated: '',
				bechdelResults: {
					pass: false,
					bechdelScore: 0,
					numScenesPass: 0,
					scenesThatPass: [],
					numScenesDontPass: 0,
					numOfFemalesChars: 0,
					numOfMaleChars: 0,
					numOfFemalesCharsWithDialogue: 0,
					numOfMaleCharsWithDialogue: 0,
					totalLinesFemaleDialogue: 0,
					totalLinesMaleDialogue: 0,
				},
			},
			loading: true,
		};

		this.getFilm = this.getFilm.bind(this);
	}

	componentDidMount() {
		try {
			this.getFilm();
		} catch (error) {
			console.log(error);
		}
	}

	getFilm() {
		const { id } = this.props.match.params;
		const url = `/api/film/${id}`;
		const options = {
			method: 'GET',
		};

		this.setState({
			loading: true,
		});

		api(url, options)
			.then(data => {
				// console.log(data);
				this.setState({
					film: data,
					loading: false,
				});
				return data;
			})
			.catch(err => {
				console.error(err);
				this.setState({
					loading: false,
				});
			});
	}

	render() {
		return <Film {...this.state} />;
	}
}

FilmContainer.propTypes = {
	match: PropTypes.shape({
		params: PropTypes.shape({
			id: PropTypes.string,
		}),
	}).isRequired,
};

FilmContainer.defaultProps = {};

export default FilmContainer;
