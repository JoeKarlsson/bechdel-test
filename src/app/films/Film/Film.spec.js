import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import Film from './Film';

jest.mock('../../helper/api');

configure({ adapter: new Adapter() });

describe('Film', () => {
	let wrapper;
	let inst;
	const router = {
		params: {
			id: 1234,
		},
	};

	beforeEach(() => {
		wrapper = shallow(<Film match={router} />);
		inst = wrapper.instance();
	});

	describe('rendering', () => {
		describe('initial state', () => {
			it('match the snapshot', () => {
				const component = renderer.create(
					<MemoryRouter>
						<Film match={router} />
					</MemoryRouter>
				);
				const tree = component.toJSON();
				expect(tree).toMatchSnapshot();
			});
			it('is rendered correctly', () => {
				expect(wrapper).toHaveLength(1);
			});
			it('should not have any inital props', () => {
				const initialProps = inst.props;
				const expectedProps = {
					match: {
						params: {
							id: 1234,
						},
					},
				};
				expect(initialProps).toMatchObject(expectedProps);
			});
			it('should not have any inital state', () => {
				const initialState = inst.state;
				const expectedState = {
					film: {
						title: '',
						images: {
							poster: '',
							backdrop: '',
						},
						plot: '',
						directors: [],
						writers: [],
						genres: [],
						rated: '',
						bechdelResults: {
							pass: 'false',
							bechdelScore: 0,
							numScenesPass: 0,
							scenesThatPass: [],
							numScenesDontPass: 0,
							numOfFemalesChars: 0,
							numOfMaleChars: 0,
							numOfFemalesCharsWithDialogue: 0,
							numOfMaleCharsWithDialogue: 0,
							totalLinesFemaleDialogue: 0,
							totalLinesMaleDialogue: 0,
						},
					},
				};
				expect(initialState).toMatchObject(expectedState);
			});
		});
	});
});
