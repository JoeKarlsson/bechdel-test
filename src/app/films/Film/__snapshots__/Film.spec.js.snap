// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Film rendering initial state should render data state correctly 1`] = `
<div
  className="filmInfo"
>
  <div
    className="film-hero"
  >
    <img
      alt="Boyhood"
      className="film-image"
      src="https://image.tmdb.org/t/p/w1000/aE1gbq6nw8zyVqvEBXMVMCqZpCs.jpg"
    />
    <span
      className="film-hero-title"
    >
      <h4>
        Boyhood
      </h4>
      <h3>
        Bechdel Pass: 
        TRUE
      </h3>
      <p>
        Bechdel Score: 
        1
         of 3
      </p>
    </span>
  </div>
  <div
    className="plot container"
  >
    <p>
      The life of Mason, from early childhood to his arrival at college.
    </p>
    <span
      className="results container"
    >
      <div
        className="filmData"
      >
        <span
          className="catName"
        >
          Directors:
        </span>
         
        <span>
          Richard Linklater
           | 
        </span>
        <span>
          John Doe
        </span>
        <br />
        <span
          className="catName"
        >
          Writers:
        </span>
         
        <span>
          Richard Linklater
           | 
        </span>
        <span>
          John Doe
        </span>
         
        <br />
        <span
          className="catName"
        >
          Genre:
        </span>
         
        <span>
          Comedy
           | 
        </span>
        <span>
          Drama
        </span>
        <br />
        <span
          className="catName"
        >
          Rated:
        </span>
         
        97
        <br />
        <span
          className="catName"
        >
          IMDB:
        </span>
         
        <a
          href="http://www.imdb.com/title/undefined"
          target="_blank"
        >
          Boyhood
        </a>
        <br />
      </div>
      <div
        className="BechdelResults"
      >
        <span
          className="catName"
        >
          Bechdel Score:
        </span>
         
        1
         of 3
        <br />
        <span
          className="catName"
        >
          Number of Scenes that pass:
        </span>
         
        45
        <br />
        <span
          className="catName"
        >
          Number of Scenes that dont pass:
        </span>
         
        123
        <br />
        <span
          className="catName"
        >
          Number Of Females Characters:
        </span>
         
        23
        <br />
        <span
          className="catName"
        >
          Number Of Male Characters:
        </span>
         
        54
        <br />
        <span
          className="catName"
        >
          Number of Females Characters With Dialogue:
        </span>
         
        12
        <br />
        <span
          className="catName"
        >
          Number of Male Characters With Dialogue:
        </span>
         
        23
        <br />
        <span
          className="catName"
        >
          Total Lines of Female Dialogue:
        </span>
         
        1012
        <br />
        <span
          className="catName"
        >
          Total Lines of Male Dialogue:
        </span>
         
        1202
        <br />
      </div>
    </span>
    <a
      href="/"
      onClick={[Function]}
    >
      <button>
        All Films
      </button>
    </a>
  </div>
</div>
`;

exports[`Film rendering initial state should render error state correctly 1`] = `
<div>
  Im sorry! Please try again.
</div>
`;

exports[`Film rendering initial state should render zero state correctly 1`] = `
<div>
  <span
    className="Loader"
  >
    <div
      className="Loader_indicator"
    >
      <h1>
        <span>
          Destroying the Patriarchy
        </span>
        <span
          className="Loader_ellipsis"
        >
          <span
            className="Loader_ellipsisDot"
          >
            .
          </span>
          <span
            className="Loader_ellipsisDot"
          >
            .
          </span>
          <span
            className="Loader_ellipsisDot"
          >
            .
          </span>
        </span>
      </h1>
    </div>
  </span>
</div>
`;
